pkgs/dart_model/lib/src/dart_model.dart

```diff
@@ -11,3 +12,65 @@ extension QualifiedNameExtension on QualifiedName {
 
   bool equals(QualifiedName other) => other.uri == uri && other.name == name;
 }
+
+extension ModelExtension on Model {
+  /// Returns the path in the model to [member], or `null` if [member] is not in this `Model`.
+  ///
+  /// TODO(davidmorgan): this works for any node, but it's not clear yet which types of
+  /// node we want this functionality exposed for.
+  /// TODO(davidmorgan): a list of path segments is probably more useful than `String`.
+  String? pathToMember(Member member) => _pathTo(member.node);
+
+  /// Returns the path in the model to [node], or `null` if [node] is not in this `Model`.
+  String? _pathTo(Map<String, Object?> node) {
+    if (node == this.node) return '';
+    final parent = _getParent(node);
+    if (parent == null) return null;
+    for (final entry in parent.entries) {
+      if (entry.value == node) {
+        final parentPath = _pathTo(parent);
+        return parentPath == null ? null : '${_pathTo(parent)}/${entry.key}';
+      }
+    }
+    return null;
+  }
+
+  /// Gets the `Map` that contains [node], or `null` if there isn't one.
+  Map<String, Object?>? _getParent(Map<String, Object?> node) {
+    // If both maps are in the same `JsonBufferBuilder` then the parent is
+    // immediately available.
+    if (this is MapInBuffer && node is MapInBuffer) {
+      final thisMapInBuffer = this as MapInBuffer;
+      final thatMapInBuffer = node as MapInBuffer;
+      if (thisMapInBuffer.buffer == thatMapInBuffer.buffer) {
+        return thatMapInBuffer.parent;
+      }
+    }
+    // Otherwise, build a `Map` of references to parents and use that.
+    return _parentsMap[node];
+  }
+
+  /// Gets a `Map` from values to parent `Map`s.
+  Map<Map<String, Object?>, Map<String, Object?>> get _parentsMap {
+    var result = _parentsMaps[this];
+    if (result == null) {
+      result =
+          _parentsMaps[this] = <Map<String, Object?>, Map<String, Object?>>{};
+      _buildParentsMap(node, result);
+    }
+    return result;
+  }
+
+  /// Builds a `Map` from values to parent `Map`s.
+  static void _buildParentsMap(Map<String, Object?> parent,
+      Map<Map<String, Object?>, Map<String, Object?>> result) {
+    for (final child in parent.values.whereType<Map<String, Object?>>()) {
+      result[child] = parent;
```

Comment:
Is it possible this already exists in some cases (I think it is?). Consider a `putIfAbsent` here which only recursively calls `_buildParentsMap` in the absent case.

Reply:
It shouldn't: every node should have a unique path, with no cycles and no reuse. Otherwise using paths like this doesn't really work.

Added a check to throw if there's a duplicate, and tests to check it throws for a cycle or a duplicate.

Reply:
The case I was thinking of is two different nodes which are siblings both ask for parents, they will both end up re-computing the parents of their parents?

Reply:
I missed replying to this yesterday, sorry about that.

Actually that's not a problem: the parents map is always computed for the whole `Model` if it's needed; then it won't be computed again. So no re-computation :)